### PR TITLE
DOCSP-35364: Flutter SDK - Nested Collections of Mixed Data

### PR DESCRIPTION
**Pull request by cbullinger**
## Pull Request Info

> **NOTE:** This PR is temporarily using an alpha snapshot - unit tests will fail. Upgrading to v2.0.0 release, once it's available, may impact some of the code examples.

Jira ticket: https://jira.mongodb.org/browse/DOCSP-35364

- [Define an Object Schema](https://preview-mongodbcbullinger.gatsbyjs.io/realm/docsp-35364-dart-collections-mixed/sdk/flutter/realm-database/model-data/define-realm-object-schema/#model-unstructured-data): Added new "Modeled Unstructured Data" section with example
- [Data Types](https://preview-mongodbcbullinger.gatsbyjs.io/realm/docsp-35364-dart-collections-mixed/sdk/flutter/realm-database/model-data/data-types/#realmvalue): Updated ``RealmValue`` mixed data type section with new info and examples for collections in mixed; added note re. breaking change (**links to migration page from a separate PR**)
- [Read](https://preview-mongodbcbullinger.gatsbyjs.io/realm/docsp-35364-dart-collections-mixed/sdk/flutter/crud/read/#query-nested-collections-of-mixed-data): Add new Query Nested Collections of Mixed Data (**needs query examples added as part of RQL ticket**)

### Reminder Checklist

Before merging your PR, make sure to check a few things.

- [x] Did you tag pages appropriately?
  - genre
  - meta.keywords
  - meta.description
- [x] Describe your PR's changes in the Release Notes section
- [x] Create a Jira ticket for related docs-app-services work, if any

### Release Notes

- **Flutter** SDK
  - Realm Database/Model Data/Define an Object Schema: New "Modeled Unstructured Data" section added with a tested, Bluehawked code example to document nested collections of mixed data. 
  - Realm Database/Model Data/Data Types: ``RealmValue`` section updated to document the new support for nested collections of mixed data and additional changes to the data type. Tested, Bluehawked code examples were updated or added.
  - Realm Database/CRUD/Read: New "Query Nested Collections of Mixed Data" section added with a tested, Bluehawked code example to document new RQL queries for nested collections. 

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt [hunk 1]
```diff
@@ -233,6 +234,65 @@ If you're using Atlas Device Sync, the name that you specify in the
...
+variable structure or data whose shape or type is not known at runtime. 
+For example, you might have highly variable user-created objects, event logs, 
+or survey response data that are collected and stored in a variety of JSON 
+formats. It could also be a good temporary option while you are 
+in early development and still determining the structure of your app's data. 
+
+To model unstructured data in your schema using collections of mixed type, 
+define the appropriate properties in your schema as :ref:`RealmValue <flutter-realm-value>` types 
+that contain either a :ref:`RealmList <flutter-realm-list>` or a 
+:ref:`RealmMap <flutter-realm-map>` of ``RealmValue`` type:
```
Comment (nirinchev) [inline source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt:260]: I know what this means, but to a new user reading this, it may seem they need to somehow clarify that the RealmValue property contains a map or list.

File: source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt [hunk 2]
```diff
@@ -233,6 +234,65 @@ If you're using Atlas Device Sync, the name that you specify in the
...
+formats. It could also be a good temporary option while you are 
+in early development and still determining the structure of your app's data. 
+
+To model unstructured data in your schema using collections of mixed type, 
+define the appropriate properties in your schema as :ref:`RealmValue <flutter-realm-value>` types 
+that contain either a :ref:`RealmList <flutter-realm-list>` or a 
+:ref:`RealmMap <flutter-realm-map>` of ``RealmValue`` type: 
+
+- ``RealmValue<List<RealmValue>>``
+- ``RealmValue<RealmMap<String, RealmValue>>``
```
Comment (nirinchev) [inline source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt:263]: This is not valid dart. The field type is a `RealmValue` with no generic arguments - it can implicitly contain a primitive or a collection with no extra definition required.

File: source/examples/generated/flutter/define_realm_model_test.snippet.create-unstructured-data-example.dart [hunk 3]
```diff
@@ -0,0 +1,24 @@
+realm.write(() {
+  final eventLog = realm.add(
+      EventLog(ObjectId(), 'purchase', DateTime.now(), 'user123', details: {
```
Comment (nirinchev) [inline source/examples/generated/flutter/define_realm_model_test.snippet.create-unstructured-data-example.dart:3]: I think we need to show the schema of the `EventLog` somewhere.

Also, if `details` is a RealmValue, you can probably move the `RealmValue.from` to the top level rather than have it in multiple places in the structure:

```dart
details: RealmValue.from({
  'ipAddress': '192.168.1.1',
  items: [ ... ]
}),
```

File: source/sdk/flutter/realm-database/model-data/data-types.txt [hunk 4]
```diff
@@ -339,39 +343,102 @@ following things are true when using ``compareTo()``:
...
+.. important:: Flutter SDK v2.0.0 Breaking Changes to ``RealmValue``
+
+   Flutter SDK version 2.0.0 updates ``RealmValue`` to allow a ``List`` 
+   or ``Map`` type of ``RealmValue``, which enables more flexibility 
+   when modeling unstructured data. Refer to  :ref:`flutter-model-unstructured-data` 
+   for more information.
+   
+   This update also includes the following breaking changes: 
+   
+   - ``RealmValue.type`` is now an enum of ``RealmValueType`` instead of ``Type``, and ``RealmValue.value.runtimeType`` is added.
```
Comment (nirinchev) [inline source/sdk/flutter/realm-database/model-data/data-types.txt:355]: `RealmValue.value.runtimeType` is not technically added. `Object.runtimeType` comes from the framework (i.e. it's built in for all dart objects). It's just that previously `type` was just a shorthand for `value.runtimeType` and now users can just access it themselves.

File: source/sdk/flutter/realm-database/model-data/data-types.txt [hunk 5]
```diff
@@ -339,39 +343,102 @@ following things are true when using ``compareTo()``:
...
+   - ``RealmValue.uint8List`` is renamed to ``RealmValue.binary``.
+
+   These changes require a migration for existing apps.
```
Comment (nirinchev) [inline source/sdk/flutter/realm-database/model-data/data-types.txt:358]: I think this may be a little extraneous information and also has the possibility to throw people off. Migration is typically associated with data migrations and might imply that you'd be forced to execute some migration on your database, when in reality, one needs to just update some code. We should also qualify the statement with "may" since there are probably many apps that don't use `RealmValue` at all and even the ones that do may not have used the `.type` property.

File: source/sdk/flutter/realm-database/model-data/data-types.txt [hunk 6]
```diff
@@ -339,39 +343,102 @@ following things are true when using ``compareTo()``:
...
+
+   For information on how to update an existing app from an earlier 
+   version to v2.0.0 or later, refer to :ref:`flutter_upgrade-v2`. 
+
 The `RealmValue <https://pub.dev/documentation/realm_common/latest/realm_common/RealmValue-class.html>`__
-data type is a mixed data type that can represent any other valid Realm data type except a collection.
-You can create collections  of type ``RealmValue``, but a ``RealmValue`` itself
-cannot be a collection. ``RealmValue`` is indexable, but cannot be a primary key.
+data type is a mixed data type that can represent any other valid Realm 
+data type. In Flutter SDK v2.0.0 and later, ``RealmValue`` can represent a List or Map collection (but *not* a Set) of mixed type.
```
Comment (nirinchev) [inline source/sdk/flutter/realm-database/model-data/data-types.txt:365]:
```suggestion
data type. In Flutter SDK v2.0.0 and later, ``RealmValue`` can represent a ``List<RealmValue>`` or ``Map<String, RealmValue>``. 
```

File: source/sdk/flutter/realm-database/model-data/data-types.txt [hunk 7]
```diff
@@ -339,39 +343,102 @@ following things are true when using ``compareTo()``:
...
+   These changes require a migration for existing apps. 
+
+   For information on how to update an existing app from an earlier 
+   version to v2.0.0 or later, refer to :ref:`flutter_upgrade-v2`. 
+
 The `RealmValue <https://pub.dev/documentation/realm_common/latest/realm_common/RealmValue-class.html>`__
-data type is a mixed data type that can represent any other valid Realm data type except a collection.
-You can create collections  of type ``RealmValue``, but a ``RealmValue`` itself
-cannot be a collection. ``RealmValue`` is indexable, but cannot be a primary key.
+data type is a mixed data type that can represent any other valid Realm
```
Comment (nirinchev) [inline source/sdk/flutter/realm-database/model-data/data-types.txt:364]: It cannot represent embedded objects.

File: source/sdk/flutter/realm-database/model-data/data-types.txt [hunk 8]
```diff
@@ -339,39 +343,102 @@ following things are true when using ``compareTo()``:
...
+
+.. versionchanged:: 2.0.0
+   ``RealmValueType`` enum replaces ``RealmValue.type``.
+   ``RealmValue.value.runtimeType`` added.
+   ``RealmValue.binary`` replaces  ``RealmValue.uint8List``.
+
+In Flutter SDK v2.0.0 and later, you access the type of data with the 
+``RealmValueType`` enum and the value with ``RealmValue.value``. 
+In earlier SDK versions, you access the type of data with 
+``RealmValue.type``.
```
Comment (nirinchev) [inline source/sdk/flutter/realm-database/model-data/data-types.txt:398]: I think this paragraph is confusing. I would suggest rewording it to something along the lines of:

> To obtain the data stored in a `RealmValue`, you can use `RealmValue.value` which returns an `Object?` or `RealmValue.as<T>` to fetch the data and cast it to a desired type. You can check the type of the data stored in a `RealmValue` by accessing the `type` property. Starting with Realm Flutter v2.0.0, that will return a `RealmValueType` enum value, while in earlier versions it would return `RealmValue.value.runtimeType`.

File: source/sdk/flutter/realm-database/model-data/data-types.txt [hunk 9]
```diff
@@ -339,39 +343,102 @@ following things are true when using ``compareTo()``:
...
 
 .. literalinclude:: /examples/generated/flutter/data_types_test.snippet.realm-value-type-value.dart
    :language: dart
 
+In version 2.0.0 and later, you can also access the runtime type with ``RealmValue.value.runtimeType``.
+
+.. literalinclude:: /examples/generated/flutter/data_types_test.snippet.realm-value-runtime-type.dart
+   :language: dart
+
+.. _flutter-nested-collections-realm-value:
```
Comment (nirinchev) [inline source/sdk/flutter/realm-database/model-data/data-types.txt:408]: I don't think we should show this - it's generally a bad practice and we only added it in the past to have something. Accessing the runtimeType has always been available, but instead of doing that, you'd probably want to use an expression:

```dart
final value = mixedValue.value;
if (value is int) {
  sum = sum + value;
} else if (value is String) {
  final subValue = value.substring(5, 8);
}
```

File: source/examples/generated/flutter/data_types_test.snippet.realm-value-type-value.dart [hunk 10]
```diff
@@ -1,6 +1,16 @@
...
-  if (obj.singleAnyValue.type == int) {
-    print(obj.singleAnyValue.value.toString());
+  switch (obj.singleAnyValue.type) {
+    // Use RealmValueType data type enums
+    case RealmValueType.int:
+      print('Int value: ${obj.singleAnyValue.value}');
+      break;
+    case RealmValueType.string:
+      print('String value: ${obj.singleAnyValue.value}');
+      break;
```
Comment (nirinchev) [inline source/examples/generated/flutter/data_types_test.snippet.realm-value-type-value.dart:10]: This is not super helpful as `obj.singleAnyValue.value` is of type `Object?` and unless you're stringifying it, you can't use it for anything meaningful without casting. We could instead showcase the `.as<T>` helper method and show some use case that is using the actual type - e.g.:

```dart
final anyValue = obj.singleAnyValue;
switch (anyValue.type) {
  case RealmValueType.int:
    // we stored birth year
    approximateAge = DateTime.now().year - anyValue.as<int>();
    break;
  case RealmValueType.date:
    // We stored the birthday
    approximateAge = (DateTime.now().difference(anyValue.as<DateTime>()).inDays / 365).floor();
    break;
  case RealmValueType.string:
    // We stored the birthday as a string
    final birthday = DateTime.parse(anyValue.as<String>());
    approximateAge = (DateTime.now().difference(birthday).inDays / 365).floor();
    break;
}
```

File: source/sdk/flutter/realm-database/model-data/data-types.txt [hunk 11]
```diff
@@ -339,39 +343,102 @@ following things are true when using ``compareTo()``:
...
+Nested Collections of RealmValue
+````````````````````````````````
+
+.. versionadded:: v2.0.0
+
+Starting in Flutter SDK version 2.0.0, a ``RealmValue`` data type can 
+contain collections (a list or map, but *not* a set) of ``RealmValue`` type: 
+
+- ``RealmValue<List<RealmValue>>``
+- ``RealmValue<RealmMap<String, RealmValue>>``
```
Comment (nirinchev) [inline source/sdk/flutter/realm-database/model-data/data-types.txt:419]: These declarations are incorrect - it's just a `RealmValue`.

File: source/examples/generated/flutter/data_types_test.snippet.realm-value-nested-collections.dart [hunk 12]
```diff
@@ -0,0 +1,35 @@
+final singleAnyValue = RealmValue.from(1);
+final listOfAnyValue = RealmValue.from([singleAnyValue]);
```
Comment (nirinchev) [inline source/examples/generated/flutter/data_types_test.snippet.realm-value-nested-collections.dart:2]: I think it's a little unclear what `singleAnyValue` is here and below.

File: source/examples/generated/flutter/data_types_test.snippet.realm-value-nested-collections.dart [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+final mapOfAnyValue = RealmValue.from({
+  '1': singleAnyValue,
+  '2': singleAnyValue,
+  '3': RealmValue.nullValue()
+});
+
+realm.write(() {
+  var collectionsOfMixed = realm.add(RealmValueExample(
```
Comment (nirinchev) [inline source/examples/generated/flutter/data_types_test.snippet.realm-value-nested-collections.dart:10]: Do we need the assignment here?

File: source/examples/generated/flutter/data_types_test.snippet.realm-value-nested-collections.dart [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+      singleAnyValue: singleAnyValue,
+      listOfMixedAnyValues: [singleAnyValue, singleAnyValue],
+      mapOfMixedAnyValues: {'key': singleAnyValue}));
```
Comment (nirinchev) [inline source/examples/generated/flutter/data_types_test.snippet.realm-value-nested-collections.dart:13]: I think these are kind of confusing - `singleAnyValue` or `listOfMixedAnyValues` are somewhat convoluted names and not really relatable to a real-world example.

File: source/examples/generated/flutter/data_types_test.snippet.realm-value-nested-collections.dart [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+      singleAnyValue: singleAnyValue,
+      listOfMixedAnyValues: [
+        RealmValue.from([
+          listOfAnyValue,
+          RealmValue.from([
+            listOfAnyValue,
+            RealmValue.from([
+              listOfAnyValue, mapOfAnyValue, singleAnyValue
+              ])
+          ]),
```
Comment (nirinchev) [inline source/examples/generated/flutter/data_types_test.snippet.realm-value-nested-collections.dart:24]: I think it'd be nicer if we could show a complex data structure inline rather than referring objects we defined above. Seeing a json-like data structure is a lot easier than to reason about what's going on where, especially with all the `RealmValue.from`-s being scattered around.

**Review by nirinchev — COMMENTED** [review]
I haven't gone through the whole thing yet since it's quite late, but I'll get to it on Monday! I think the structure is sound, but we could come up with better examples. There are a few inaccuracies as well that I've highlighted.

File: source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt (hunk 1 above)
Comment (cbullinger) [reply source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt:260]: Updated for clarity

File: source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt (hunk 2 above)
Comment (cbullinger) [reply source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt:263]: Updated here and in the Data Types page

File: source/sdk/flutter/realm-database/model-data/data-types.txt (hunk 9 above)
Comment (cbullinger) [reply source/sdk/flutter/realm-database/model-data/data-types.txt:408]: Replaced this with example of accessing the type with an expression

File: source/sdk/flutter/realm-database/model-data/data-types.txt (hunk 8 above)
Comment (cbullinger) [reply source/sdk/flutter/realm-database/model-data/data-types.txt:398]: Updated this and the example to demonstrate both methods of accessing data

File: source/sdk/flutter/realm-database/model-data/data-types.txt (hunk 11 above)
Comment (cbullinger) [reply source/sdk/flutter/realm-database/model-data/data-types.txt:419]: Updated here and in Define Schema page

File: source/examples/generated/flutter/data_types_test.snippet.realm-value-nested-collections.dart (hunk 12 above)
Comment (cbullinger) [reply source/examples/generated/flutter/data_types_test.snippet.realm-value-nested-collections.dart:2]: let's sync up on a better example to address this and the following comments

File: source/examples/generated/flutter/data_types_test.snippet.realm-value-type-value.dart (hunk 10 above)
Comment (cbullinger) [reply source/examples/generated/flutter/data_types_test.snippet.realm-value-type-value.dart:10]: Updated

File: source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt [hunk 16]
```diff
@@ -233,6 +234,63 @@ If you're using Atlas Device Sync, the name that you specify in the
...
+within a  ``RealmValue`` property.
+
+The ability to nest collections of mixed data enables you to define data 
+that doesn't otherwise conform to an expected schema, including data with 
+variable structure or data whose shape or type is not known at runtime. 
+For example, you might have highly variable user-created objects, event logs, 
+or survey response data that are collected and stored in a variety of JSON 
+formats. This approach allows you to :ref:`react to changes <flutter-realm-list-change-listener>`
+in the nested data and to update specific elements, but it is less 
+performant than using a structured schema or serializing JSON blobs
```
Comment (lindseymoore) [inline source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt:256]:
```suggestion
in the nested data and to update specific elements. However in some cases, it is less 
performant than using a structured schema or serializing JSON blobs.
```
Why it is less performant? Maybe give some reasons or cases where it would be less performant.

File: source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt [hunk 17]
```diff
@@ -233,6 +234,63 @@ If you're using Atlas Device Sync, the name that you specify in the
...
+Note that ``RealmValue`` *cannot* represent a ``RealmSet`` or an embedded object.
+
+For example, you might use a ``RealmValue`` that contains a map of mixed 
+data when modeling a variable event log object: 
+
+.. literalinclude:: /examples/generated/flutter/define_realm_model_test.snippet.unstructured-data-model.dart
+   :language: dart
+
+.. io-code-block::
+
```
Comment (lindseymoore) [inline source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt:273]:
```suggestion
    :copyable: true
```
Did a quick search, adding this should make the code snippet copyable!

**Review by lindseymoore — CHANGES_REQUESTED** [review]
Overall, content looks good! A few smaller suggestions and fixes. 
For code snippets, I would like to see introduction comments to explain what the code is doing. Sometimes you explain what the code is doing in the text, sometimes not. I think an introduction comment is helpful whether the code is explained in text or not, given the code is copyable.

File: source/sdk/flutter/realm-database/model-data/data-types.txt [hunk 18]
```diff
@@ -339,30 +343,96 @@ following things are true when using ``compareTo()``:
...
+   for more information.
+   
+   This update also includes the following breaking changes, which may affect 
+   your app when updating to v2.0.0 or later: 
+   
+   - ``RealmValue.type`` is now an enum of ``RealmValueType`` instead of ``Type``.
+   - ``RealmValue.uint8List`` is renamed to ``RealmValue.binary``. 
+
+   For more information on how to update an existing app from an earlier 
+   version to v2.0.0 or later, refer to :ref:`flutter-upgrade-v2`.
```
[the remark this answers is not in the document]
Comment (cbullinger) [reply source/sdk/flutter/realm-database/model-data/data-types.txt:360]: not yet, the page is part of another PR. i've commented these last lines out for now, and will re-add it once the page is available

File: source/sdk/flutter/crud/read.txt [hunk 19]
```diff
@@ -204,6 +204,24 @@ or primitives.
...
+Query Nested Collections of Mixed Data
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. versionadded:: 2.0.0
+
+In Flutter SDK v2.0.0 and later, :ref:`RealmValue <flutter-realm-value>` 
+properties can contain collections (a list or map) of mixed data. These 
+collections can be nested within collections and can contain other 
+collections of mixed data. You can query these with dot notation, the 
+same way you would a normal collection. You can also use bracket notation to query a specific element in the collection (for example, ``[FIRST]``, ``[1]``, ``[LAST]``, or ``[*]``).
```
[the remark this answers is not in the document]
Comment (cbullinger) [reply source/sdk/flutter/crud/read.txt:218]: Was going to rely on the RQL docs to provide the bulk of this, but I will update these with more information since the RQL content isn't ready yet

File: source/sdk/flutter/realm-database/model-data/data-types.txt [hunk 20]
```diff
@@ -339,30 +343,96 @@ following things are true when using ``compareTo()``:
...
 .. literalinclude:: /examples/generated/flutter/data_types_test.snippet.realm-value-from.dart
    :language: dart
 
-Access the type of data with ``RealmValue.type`` and the value with ``RealmValue.value``.
+.. _flutter-access-realm-value-type:
+
+Access RealmValue Data Type 
+```````````````````````````
+
+.. versionchanged:: 2.0.0
```
[the remark this answers is not in the document]
Comment (cbullinger) [reply source/sdk/flutter/realm-database/model-data/data-types.txt:391]: i think it's just how the description portion of the versionchanged direction is formatted, unfortunately...

File: source/sdk/flutter/realm-database/model-data/data-types.txt [hunk 21]
```diff
@@ -339,30 +343,96 @@ following things are true when using ``compareTo()``:
...
+.. literalinclude:: /examples/generated/flutter/data_types_test.snippet.realm-value-value.dart
+   :language: dart
+   :emphasize-lines: 3, 11
+
+You can check the type of data currently stored in a ``RealmValue`` property by 
+accessing the type property. Starting with Flutter SDK v2.0.0, this returns a 
+``RealmValueType`` enum. In earlier SDK versions, the SDK returned a 
+``RealmValue.value.runtimeType``.
+
+.. literalinclude:: /examples/generated/flutter/data_types_test.snippet.realm-value-type.dart
```
[the remark this answers is not in the document]
Comment (cbullinger) [reply source/sdk/flutter/realm-database/model-data/data-types.txt:409]: Updated snippet and highlighting

File: source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt (hunk 16 above)
Comment (cbullinger) [reply source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt:256]: I've noted this for a fast follow addition. I'll need to get more info from engineering for the additional details

File: source/sdk/flutter/crud/read.txt [hunk 22]
```diff
@@ -204,6 +204,39 @@ or primitives.
...
+collections can be nested within collections and can contain other 
+collections of mixed data. 
+
+You can query these using the same RQL operators and syntax as you would for a  
+normal list or dictionary. Refer to the :ref:`rql` documentation for more
+information.  
+
+For nested collections, you can also use: 
+
+- Bracket notation, which provides additional collection query support:
```
Comment (lindseymoore) [inline source/sdk/flutter/crud/read.txt:225]: Not sure why it's bolded in the staging link, but remove bolding if possible.

File: source/sdk/flutter/crud/read.txt [hunk 23]
```diff
@@ -204,6 +204,39 @@ or primitives.
...
+information.  
+
+For nested collections, you can also use: 
+
+- Bracket notation, which provides additional collection query support: 
+    - ``[FIRST]`` and [``LAST``] - match the first or last elements within the collection
+    - ``[<int>]`` - match the element at the specific index 
+    - ``[*]`` - match any elements within the collection
+    - ``[SIZE]`` - match the collection length 
+- The ``@type`` operator, which has the following possible values:
```
Comment (lindseymoore) [inline source/sdk/flutter/crud/read.txt:232]: Remove bolding here too.

File: source/sdk/flutter/crud/read.txt [hunk 24]
```diff
@@ -204,6 +204,39 @@ or primitives.
...
+collections of mixed data. 
+
+You can query these using the same RQL operators and syntax as you would for a  
+normal list or dictionary. Refer to the :ref:`rql` documentation for more
+information.  
+
+For nested collections, you can also use: 
+
+- Bracket notation, which provides additional collection query support: 
+    - ``[FIRST]`` and [``LAST``] - match the first or last elements within the collection
```
Comment (lindseymoore) [inline source/sdk/flutter/crud/read.txt]:
```suggestion
    - ``[FIRST]`` and [``LAST``] : match the first or last elements within the collection
```
Nit: The hyphens for the few following lines should be colons according to the style guide: https://www.mongodb.com/docs/meta/style-guide/style/punctuation/#colons

File: source/sdk/flutter/crud/read.txt [hunk 25]
```diff
@@ -204,6 +204,42 @@ or primitives.
...
+
+You can query these using the same RQL operators and syntax as you would for a  
+normal list or dictionary. Refer to the :ref:`rql` documentation for more
+information.  
+
+For nested collections, you can also use: 
+
+- Bracket notation, which provides additional collection query support: 
+
+  - ``[FIRST]`` and [``LAST``] - match the first or last elements within the collection
```
Comment (lindseymoore) [inline source/sdk/flutter/crud/read.txt:227]:
```suggestion
  - ``[FIRST]`` and [``LAST``]: match the first or last elements within the collection
```
Nit: This should be a colon according to style guide: https://www.mongodb.com/docs/meta/style-guide/style/punctuation/#colons

File: source/sdk/flutter/crud/read.txt [hunk 26]
```diff
@@ -204,6 +204,42 @@ or primitives.
...
+  - ``[*]`` - match any elements within the collection
+  - ``[SIZE]`` - match the collection length 
+
+- The ``@type`` operator, which has the following possible values: 
+
+  - ``array`` and ``list`` - match a list collection
+  - ``dictionary`` and ``object`` - match a map collection
+  - ``collection`` - matches a list or a map collection  
+
+.. literalinclude:: /examples/generated/flutter/read_write_data_test.snippet.filter-nested-collections.dart
```
Comment (lindseymoore) [inline source/sdk/flutter/crud/read.txt:238]: Would be nice to add a comment introducing the object models since the relevant information to the section starts so far down.

File: source/sdk/flutter/realm-database/model-data/data-types.txt [hunk 27]
```diff
@@ -339,30 +343,106 @@ following things are true when using ``compareTo()``:
...
+data type except embedded objects. In Flutter SDK v2.0.0 and later, ``RealmValue`` 
+can represent a ``List<RealmValue>`` or ``Map<String, RealmValue>``. 
+
+``RealmValue`` is indexable, but cannot be a primary key. You can also create 
+collections of type ``RealmValue``. 
+
+Define a RealmValue Property
+````````````````````````````
+
+To define a property as ``RealmValue``, set its type in your object model.
```
Comment (lindseymoore) [inline source/sdk/flutter/realm-database/model-data/data-types.txt:374]:
```suggestion
To define a ``RealmValue`` property , set its type in your object model. 
```

File: source/sdk/flutter/realm-database/model-data/data-types.txt [hunk 28]
```diff
@@ -339,30 +343,106 @@ following things are true when using ``compareTo()``:
...
    you can use the special ``RealmValue.nullValue()`` property.
 
-To define a property as ``RealmValue``, set its type in your Realm object model.
+Create a RealmValue 
+```````````````````
 
-.. literalinclude:: /examples/generated/flutter/data_types_test.snippet.realm-value-model.dart
+To add a ``RealmValue`` to a Realm object, call ``RealmValue.from()`` on the data or ``RealmValue.nullValue()`` to set a null value.
+
+.. literalinclude:: /examples/generated/flutter/data_types_test.snippet.realm-value-from.dart
```
Comment (lindseymoore) [inline source/sdk/flutter/realm-database/model-data/data-types.txt:390]: Suggestion: I think for readability, comments above each of the examples would be more helpful than having to read the camel case name to know what the example is for. A simple format like``// To add a single value`` would be helpful

File: source/sdk/flutter/realm-database/model-data/data-types.txt [hunk 29]
```diff
@@ -339,30 +343,106 @@ following things are true when using ``compareTo()``:
...
+- ``RealmValue.as<T>``, which fetches the data and casts it to a desired type.
+
+.. literalinclude:: /examples/generated/flutter/data_types_test.snippet.realm-value-value.dart
+   :language: dart
+   :emphasize-lines: 4, 13
+
+You can check the type of data currently stored in a ``RealmValue`` property by 
+accessing the ``type`` property. Starting with Flutter SDK v2.0.0, this returns a 
+``RealmValueType`` enum. In earlier SDK versions, the SDK returned a 
+``RealmValue.value.runtimeType``.
```
Comment (lindseymoore) [inline source/sdk/flutter/realm-database/model-data/data-types.txt:417]:
```suggestion
``RealmValue.value.runtimeType``. The following example uses RealmValueType to run calculations based on the data type.
```
Could intro example with something like that^

File: source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt [hunk 30]
```diff
@@ -233,6 +234,66 @@ If you're using Atlas Device Sync, the name that you specify in the
...
+define the appropriate properties in your schema as 
+:ref:`RealmValue <flutter-realm-value>` types. You can then set these   
+``RealmValue`` properties as a :ref:`RealmList <flutter-realm-list>` or a 
+:ref:`RealmMap <flutter-realm-map>` collection of ``RealmValue`` elements. 
+Note that ``RealmValue`` *cannot* represent a ``RealmSet`` or an embedded object.
+
+For example, you might use a ``RealmValue`` that contains a map of mixed 
+data when modeling a variable event log object: 
+
+.. literalinclude:: /examples/generated/flutter/define_realm_model_test.snippet.unstructured-data-model.dart
```
Comment (lindseymoore) [inline source/sdk/flutter/realm-database/model-data/define-realm-object-schema.txt:269]: Could add comment at the top.

**Review by lindseymoore — APPROVED** [review]
Some typos and more stuff about comments, add at your discretion. Otherwise, LGTM!

File: source/sdk/flutter/crud/read.txt (hunk 24 above)
Comment (cbullinger) [reply source/sdk/flutter/crud/read.txt]: good catch. thanks!

File: source/sdk/flutter/realm-database/model-data/data-types.txt [hunk 31]
```diff
@@ -339,30 +343,96 @@ following things are true when using ``compareTo()``:
...
+collections. They can also be :ref:`listened to for changes like a 
+normal collection <flutter-realm-list-change-listener>`.
+
+You can leverage these nested collections to define unstructured 
+or variable data in your data model (for more information, refer to 
+:ref:`<flutter-model-unstructured-data>`). 
+
+To create nested collections of mixed data in your app, define the mixed type 
+property in your data model the same way you would any other ``RealmValue`` type. 
+Then, you can create the list or map collections using ``RealmValue.from()``.
```
[the remark this answers is not in the document]
Comment (cbullinger) [reply source/sdk/flutter/realm-database/model-data/data-types.txt:443]: Added link

File: source/sdk/flutter/crud/read.txt (hunk 22 above)
Comment (cbullinger) [reply source/sdk/flutter/crud/read.txt:225]: It's a snooty thing based on the indentation and spacing. I can never remember the formatting for a regular sub-list 😒 but it's fixed now. Thanks for catching.

File: source/sdk/flutter/crud/read.txt (hunk 23 above)
Comment (cbullinger) [reply source/sdk/flutter/crud/read.txt:232]: fixed

File: source/sdk/flutter/crud/read.txt (hunk 26 above)
Comment (cbullinger) [reply source/sdk/flutter/crud/read.txt:238]: Added. Good call.

File: source/sdk/flutter/realm-database/model-data/data-types.txt (hunk 27 above)
Comment (cbullinger) [reply source/sdk/flutter/realm-database/model-data/data-types.txt:374]: 😅 thanks for catching

File: source/sdk/flutter/realm-database/model-data/data-types.txt (hunk 29 above)
Comment (cbullinger) [reply source/sdk/flutter/realm-database/model-data/data-types.txt:417]: added
